go/enclave/components/rollup_producer.go

```diff
@@ -0,0 +1,155 @@
+package components
+
+// todo (@stefan) - once the cross chain messages based bridge is implemented remove this completely
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	gethlog "github.com/ethereum/go-ethereum/log"
+
+	"github.com/obscuronet/go-obscuro/contracts/generated/MessageBus"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/errutil"
+
+	"github.com/obscuronet/go-obscuro/go/enclave/crypto"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/obscuronet/go-obscuro/go/enclave/core"
+	"github.com/obscuronet/go-obscuro/go/enclave/db"
+)
+
+// rollupProducer encapsulates the logic of decoding rollup transactions submitted to the L1 and resolving them
+// to rollups that the enclave can process.
+type rollupProducer struct {
+	// TransactionBlobCrypto- This contains the required properties to encrypt rollups.
+	TransactionBlobCrypto crypto.TransactionBlobCrypto
+
+	ObscuroChainID  int64
+	EthereumChainID int64
+
+	logger gethlog.Logger
+
+	storage db.Storage
+
+	registry      BatchRegistry
+	blockConsumer L1BlockProcessor
+}
+
+func NewRollupProducer(
+	transactionBlobCrypto crypto.TransactionBlobCrypto,
+	obscuroChainID int64,
+	ethereumChainID int64,
+	storage db.Storage,
+	registry BatchRegistry,
+	blockConsumer L1BlockProcessor,
+	logger gethlog.Logger,
+) RollupProducer {
+	return &rollupProducer{
+		TransactionBlobCrypto: transactionBlobCrypto,
+		ObscuroChainID:        obscuroChainID,
+		EthereumChainID:       ethereumChainID,
+		logger:                logger,
+		registry:              registry,
+		blockConsumer:         blockConsumer,
+		storage:               storage,
+	}
+}
+
+// fetchLatestRollup - Will pull the latest rollup based on the current head block from the database or return null
+func (re *rollupProducer) fetchLatestRollup() (*core.Rollup, error) {
+	b, err := re.blockConsumer.GetHead()
+	if err != nil {
+		return nil, err
+	}
+	return re.getLatestRollupBeforeBlock(b)
+}
+
+func (re *rollupProducer) CreateRollup() (*core.Rollup, error) {
```

Comment:
let's add a comment here with the problem we discussed around reorgs.
That is is the responsibility of the host to wait until it knows that the rollup it has contains only "final" batches that it pushes it to the l1